### PR TITLE
feat: Add example for custom rlpx subprotocol

### LINKED ISSUE #7130: Add example crate for custom rlpx subprotocol
### Describe the feature

Add a new example that showcases rlpx subprotocols, similar to this simple Ping/Pong protocol

https://github.com/paradigmxyz/reth/blob/c9c269b64f43932fccc45cebd156d0a091e22103/crates/net/network/tests/it/multiplex.rs#L25-L26

but something more sophisticated

A protocol can be installed via:

https://github.com/paradigmxyz/reth/blob/c9c269b64f43932fccc45cebd156d0a091e22103/crates/net/network/src/network.rs#L210-L214

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by owanikin**
I am creating this draft pr to solicit feedback on:  1) If I am on the right track. 2) pointers/direction on how to really solve the issue. Thanks for your help.

Closes #7130

### CODE REVIEW

**Review by onbjerg — COMMENTED** [review]
just as a quick side note, pls remember to add the example to `examples/readme.md`

File: examples/custom-rlpx-subprotocol/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+use std::{
+    net::SocketAddr,
+    pin::Pin,
+    task::{ready, Context, Poll},
+};
+use tokio::sync::{mpsc, oneshot};
+use tokio_stream::wrappers::UnboundedReceiverStream;
+
+// Custom Rlpx Subprotocol
+pub mod proto {
```
Comment (mattsse) [inline examples/custom-rlpx-subprotocol/src/main.rs:31]: can we move this to a separate file?

File: examples/custom-rlpx-subprotocol/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+        
+
+        
+        // After lauch and after launch we inject a new rlpx protocol handler via the network node.network
+        // the rlpx can be similar to the test example, could even be something like simple string message exchange
+        
+    
+
+        // let custom_rlpx_handler = CustomRlpxProtoHandler{ state: ProtocolState { events: node.network.events.clone()}};
+        // node.network.add_rlpx_sub_protocol(custom_rlpx_handler);
```
Comment (mattsse) [inline examples/custom-rlpx-subprotocol/src/main.rs:280]: we can uncomment this

**Review by mattsse — CHANGES_REQUESTED** [review]
great, this is simple yet illustrates the feature perfectly

a few nits

File: examples/custom-rlpx-subprotocol/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+}
+
+enum Command {
+    /// Send a custom message to the peer
+    CustomMessage {
+        msg: String,
+        /// The response will be sent to this channel.
+        response: oneshot::Sender<String>,
+    },
+}
```
Comment (gakonst) [inline examples/custom-rlpx-subprotocol/src/main.rs:177]: @mattsse this is cool -- we should demonstrate a bridge/oracle on this, and eventually build up to a full consensus example i feel?

and this is all subprotocols for commands post-establishing a connection, but what about discv? how should we do discovery of nodes that support custom subprotocols?

**Comment by mattsse** [comment]
@owanikin I rebased this, but this still needs the required protocol trait implementations

**Review by gakonst — COMMENTED** [review]
bump -- do we want to get this over the line?

**Comment by loocapro** [comment]
Hi @owanikin are you still on this? I d like to get it.

**Comment by owanikin** [comment]
> Hi @owanikin are you still on this? I d like to get it.

No @Luca, please do have it.

**Comment by shekhirin** [comment]
Superseded by https://github.com/paradigmxyz/reth/pull/8991, thanks for starting the work @owanikin!

**Pull request closed without merging**